consul-lambda-registrator/extension/extension.go

```diff
@@ -157,92 +199,67 @@ func (ext *Extension) startProxy(ctx context.Context, errChan chan error) error
-	if extData.Equals(ext.data) {
-		return nil
-	}
-
 	ext.Logger.Info("starting proxy server")
 
-	// Parse the configured list of upstreams. The upstreams are configured as part of the environment
-	// so we only do this on the first time through.
-	if len(ext.upstreams) == 0 {
-		ext.upstreams, err = ext.parseUpstreams()
-		if err != nil {
-			return fmt.Errorf(errFmt, err)
-		}
-	}
-
 	// Create a proxy listener configuration for each upstream.
 	proxyConfigs := make([]*proxy.Config, len(ext.upstreams))
 	for i, upstream := range ext.upstreams {
 		// Create the listener config.
-		cfg, err := ext.proxyConfig(upstream, extData)
-		if err != nil {
-			return fmt.Errorf(errFmt, err)
-		}
-		proxyConfigs[i] = cfg
+		ext.Logger.Debug("configuring upstream", "name", upstream.Name, "port", upstream.Port)
+		proxyConfigs[i] = ext.proxyConfig(upstream)
 	}
 
-	// Drain and close the existing proxy.
-	ext.closeProxy()
-
-	// Create the proxy server.
+	// Create and start the proxy server.
 	ext.proxy = proxy.New(ext.Logger, proxyConfigs...)
-
 	go func(errChan chan error) {
-		// Get the lock for the proxy mutex. This go routine will exit and release
-		// the lock when the proxy's Close method is called. The lock ensures that
-		// we never attempt to start a new proxy server until the old one has exited.
-		ext.proxyMutex.Lock()
-		defer ext.proxyMutex.Unlock()
+		defer ext.proxy.Close()
 		errChan <- ext.proxy.Serve()
 	}(errChan)
 
-	return nil
-}
-
-func (ext *Extension) closeProxy() {
-	trace.Enter()
-	defer trace.Exit()
-
-	if ext.proxy != nil {
-		ext.proxy.Close()
+	// Wait for the proxy to be ready to serve requests before returning.
+	timeout := time.NewTicker(ext.ProxyTimeout)
+	defer timeout.Stop()
+	select {
+	case <-ctx.Done():
+		ext.Logger.Info("context closed while waiting for proxy to start.")
+	case <-ext.proxy.Wait():
+		ext.Logger.Info("proxy server ready")
+	case <-timeout.C:
+		return fmt.Errorf("timeout waiting for proxy to start")
 	}
+
+	return nil
 }
 
-func (ext *Extension) proxyConfig(upstream structs.Service, extData structs.ExtensionData) (*proxy.Config, error) {
+func (ext *Extension) proxyConfig(upstream structs.Service) *proxy.Config {
 	trace.Enter()
 	defer trace.Exit()
 
 	cfg := &proxy.Config{}
 
-	roots := x509.NewCertPool()
-	roots.AppendCertsFromPEM([]byte(extData.RootCertPEM))
-
-	cert, err := tls.X509KeyPair([]byte(extData.CertPEM), []byte(extData.PrivateKeyPEM))
-	if err != nil {
-		return cfg, err
-	}
-
 	// Listen on the upstream's port on all interfaces.
 	cfg.ListenFunc = func() (net.Listener, error) {
 		return net.Listen("tcp", fmt.Sprintf(":%d", upstream.Port))
 	}
 
 	// Wrap the outgoing request in an mTLS session and dial the mesh gateway.
 	cfg.DialFunc = func() (net.Conn, error) {
+		// Get the lock for the extension data to ensure that this func picks up the
+		// latest config. This also ensures that the extension data doesn't get updated
+		// while we are dialing out.
+		ext.dataMutex.Lock()
```

Comment:
If we make `ext.dataMutex` a `RWMutex`, will it avoid requests locking other ones?

Reply:
I don't think that it really makes a difference in the current code because there are only two places where this mutex is used. That being said, I love this suggestion because it is more correct and conveys the intent of the code in a more accurate way: The `getExtensionData` func is the writer and everyone else is a reader.. We only need to write lock when the data is being updated; otherwise, a read lock is sufficient.

Reply:
Actually you're right that we do need to use `RWMutex` to make this truely concurrent. If we don't, then multiple calls to dial out will have to wait on each other.